plugins/bnr-dp/fu-bnr-dp-firmware.c

```diff
@@ -0,0 +1,595 @@
+				    "firmware file is missing XML header");
+		return FALSE;
+	}
+
+	/* find the index of the first null byte, indicating the end of the XML header */
+	if (!fu_input_stream_find(stream,
+				  header_separator,
+				  sizeof(header_separator),
+				  &separator_idx,
+				  error))
+		return FALSE;
+
+	/* read XML header */
+	header = fu_input_stream_read_bytes(stream, 0, separator_idx, NULL, error);
+	if (header == NULL)
+		return FALSE;
+
+	if (!xb_builder_source_load_bytes(builder_source,
+					  header,
+					  XB_BUILDER_SOURCE_FLAG_NONE,
+					  error))
+		return FALSE;
+
+	xb_builder_import_source(builder, builder_source);
+	silo = xb_builder_compile(builder, XB_BUILDER_COMPILE_FLAG_SINGLE_ROOT, NULL, error);
+	if (silo == NULL)
+		return FALSE;
+
+	if (!fu_bnr_dp_firmware_header_parse(self, silo, error))
+		return FALSE;
+
+	if (!fu_bnr_dp_firmware_payload_parse(self, stream, separator_idx + 1, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+/* set FuBnrDpFirmware private data to information from device */
+gboolean
+fu_bnr_dp_firmware_parse_from_device(FuBnrDpFirmware *self,
+				     const FuStructBnrDpFactoryData *factory_data,
+				     const FuStructBnrDpPayloadHeader *fw_header,
+				     GError **error)
+{
+	g_autoptr(GBytes) bytes = NULL;
+	guint64 version = 0;
+
+	bytes = fu_firmware_get_bytes_with_patches(FU_FIRMWARE(self), error);
+	if (bytes == NULL)
+		return FALSE;
+
+	self->device_id = fu_bnr_dp_effective_product_num(factory_data);
+	self->usage = g_strdup("fw");
+	self->function = '_';
+	self->variant = fu_bnr_dp_effective_compat_id(factory_data);
+	self->payload_length = g_bytes_get_size(bytes);
+	self->payload_checksum =
+	    fu_bnr_dp_firmware_buf_checksum(g_bytes_get_data(bytes, NULL), g_bytes_get_size(bytes));
+	self->material = fu_struct_bnr_dp_factory_data_get_identification(factory_data);
+
+	if (!fu_bnr_dp_version_from_header(fw_header, &version, error))
+		return FALSE;
+	fu_firmware_set_version_raw(FU_FIRMWARE(self), version);
+
+	return TRUE;
+}
+
+static GByteArray *
+fu_bnr_dp_firmware_write(FuFirmware *firmware, GError **error)
+{
+	FuBnrDpFirmware *self = FU_BNR_DP_FIRMWARE(firmware);
+	g_autoptr(GByteArray) r = g_byte_array_new();
+	g_autoptr(GBytes) payload = NULL;
+	g_autoptr(XbBuilderNode) bn = NULL;
+	g_autofree gchar *xml = NULL;
+
+	g_autofree gchar *device_id = g_strdup_printf("%" G_GUINT64_FORMAT, self->device_id);
+	g_autofree gchar *version =
+	    g_strdup_printf("%" G_GUINT64_FORMAT, fu_firmware_get_version_raw(firmware));
+	g_autofree gchar *function = g_strdup_printf("%c", self->function);
+	g_autofree gchar *variant = g_strdup_printf("%" G_GUINT64_FORMAT, self->variant);
+	g_autofree gchar *payload_length =
+	    g_strdup_printf("%" G_GUINT64_FORMAT, self->payload_length);
+	g_autofree gchar *payload_checksum = g_strdup_printf("0x%X", self->payload_checksum);
+	g_autoptr(GDateTime) now = g_date_time_new_now_local();
+	g_autofree gchar *date = g_date_time_format(now, "%d.%m.%Y");
```

Comment:
I'd really like a simple self test for this code; e.g. https://github.com/fwupd/fwupd/blob/main/plugins/uf2/fu-self-test.c